### ISSUE TITLE
Collapse main menu options in a hamburger menu in small screens and high zoom

### ISSUE THREAD

**Issue by steff456** [jupyterlab/jupyterlab#13148]
### Problem

Currently, when Jupyterlab is opened in a small screen or the user increases the zoom the main menu options start to hide, making  the right most menus not usable,

![image](https://user-images.githubusercontent.com/20992645/192878361-8eef20ae-eba1-4cf8-8dd2-a3a91f81c7aa.png)

⬆️ `Help` menu is not visible and `Settings` menu is cropped.

### Proposed Solution

The proposed solution is to add a hamburger menu in the right side to store the menus that are not visible in such way that the user will be able to use them.

### Additional context

I think that this change will probably have to be done in Lumino, but any feedback in the proposed solution will be great!

**Comment by fcollonval** [comment]
I'm wondering if instead of displaying an hamburger, we should display an ellipsis that display the hidden entries. Something like this (taken from VS Code):

![image](https://user-images.githubusercontent.com/8435071/192982050-11efb541-49ab-4ba6-9183-5a4cbf85ce5b.png)

This comes from the following considerations:

- If the width is slightly too small to fit the menu. We will end up with a large unused empty white space.
- A single hamburger icon will feel weird in simple mode and in notebook v7:

| simple mode | notebook v7 |
| --- | --- |
| ![](https://raw.githubusercontent.com/jupyterlab/jupyterlab/master/galata/test/jupyterlab/general.test.ts-snapshots/simple-mode-jupyterlab-linux.png) | ![](https://raw.githubusercontent.com/jupyter/notebook/main/ui-tests/test/general.spec.ts-snapshots/notebook-chromium-linux.png) |

**Comment by krassowski** [comment]
Reopening as it does not seem to work yet, maybe some follow up action is needed.

**Comment by krassowski** [comment]
> I spent a few hours tonight digging into the overflow menu ("hamburger"). It's actually [turned on by default in Lumino](https://github.com/jupyterlab/lumino/blob/7f4bd513004d2aea3e96623210d015a834a4fca7/packages/widgets/src/menubar.ts#L53) so there's nothing to do to turn it on in the current build of JupyterLab.
> 
> So then I did some debugging. I discovered that the overflow implementation caches the size of each item in the menu bar. But it does this when the app loads and when these items have width 0. So the logic that shunts menu items into the overflow if the menu bar is too big for the available screen width never kicks in because it thinks the whole menu bar has zero width. 
> 
> So I fixed that bug, but then I uncovered another. And then that one, and then another. 
>
> This takes me back to when I was reviewing that diff. I asked for the code to be broken up into smaller units and for each of those units to be rigorously unit tested because the functionality was proving to be very difficult to get right. But that never happened. And each time I tested that PR, I was always able to put the menu bar into an erroneous state, by doing things like resizing the window rapidly and frequently. So I don't really know what to do.
>
> _Originally posted by @gabalafou in https://github.com/jupyterlab/jupyterlab/issues/14766#issuecomment-1857211960_

**Comment by krassowski** [comment]
To be fair this is difficult logic to get right. We feel the pain with toolbar overflow collapsing too. In particular a recent PR fixed some issues but introduced others. I wonder if we could follow @gabalafou suggestion by creating a generic well unit-tested implementation that would be used for both menubar (in lumino) and the toolbars (in jupyterlab).
